src/Ocelot/RateLimiting/RateLimiting.cs

```diff
@@ -0,0 +1,195 @@
+﻿using Microsoft.AspNetCore.Http;
+using Ocelot.Configuration;
+using System.Globalization;
+using System.Security.Cryptography;
+
+namespace Ocelot.RateLimiting;
+
+public class RateLimiting : IRateLimiting
+{
+    private readonly IRateLimitStorage _storage;
+    private static readonly object ProcessLocker = new();
+
+    public RateLimiting(IRateLimitStorage storage)
+    {
+        _storage = storage;
+    }
+
+    public virtual RateLimitCounter ProcessRequest(ClientRequestIdentity identity, RateLimitOptions options)
+    {
+        RateLimitCounter counter;
+        var rule = options.RateLimitRule;
+        var counterId = GetStorageKey(identity, options);
+
+        // Serial reads and writes
+        lock (ProcessLocker)
```

Comment:
I'm wondering why you are using lock only for the lines 26-29, why not 26-49? MemoryCache is thread safe yes, but the implemented logic isn't.

Reply:
Correct! I plan(ed) to undertake more refactoring in the next PR to achieve 100% coverage for the feature, as the current PR resolves the reported bug. However, testing the feature was challenging due to the absence of unit tests and DI interfaces, this is the main reason of refactoring. Since you've highlighted the issue with the locking block, let's address this in the current PR.

Reply:
Fixed in commit https://github.com/ThreeMammals/Ocelot/pull/1592/commits/ebd24a1180189d1ea496fe7e6896a9fc9ba5838e ✔️